optuna/samplers/_tpe/parzen_estimator.py

```diff
@@ -18,114 +33,408 @@ class _ParzenEstimatorParameters(
-                sorted_mus[prior_pos] = prior_mu
-                sorted_mus[prior_pos + 1 :] = ordered_mus[prior_pos:]
+            shape = (n_observations + 1, len(choices))
+            assert prior_weight is not None
+            value = prior_weight / (n_observations + 1)
         else:
-            order = numpy.argsort(mus)
-            # We decide the mus.
-            low_sorted_mus_high = numpy.zeros(len(mus) + 2)
-            sorted_mus = low_sorted_mus_high[1:-1]
-            sorted_mus[:] = mus[order]
-
-        # We decide the sigma.
-        if mus.size > 0:
-            low_sorted_mus_high[-1] = high
-            low_sorted_mus_high[0] = low
-            sigma = numpy.maximum(
-                low_sorted_mus_high[1:-1] - low_sorted_mus_high[0:-2],
-                low_sorted_mus_high[2:] - low_sorted_mus_high[1:-1],
-            )
-            if not consider_endpoints and low_sorted_mus_high.size > 2:
-                sigma[0] = low_sorted_mus_high[2] - low_sorted_mus_high[1]
-                sigma[-1] = low_sorted_mus_high[-2] - low_sorted_mus_high[-3]
+            shape = (n_observations, len(choices))
+            assert prior_weight is not None
+            value = prior_weight / n_observations
+        weights = np.full(shape, fill_value=value)
+        weights[np.arange(n_observations), observations] += 1
+        weights /= weights.sum(axis=1, keepdims=True)
+        return weights
+
+    def _calculate_numerical_params(
+        self, observations: np.ndarray, param_name: str
+    ) -> Tuple[np.ndarray, np.ndarray]:
+
+        n_observations = self._n_observations
+        consider_prior = self._parameters.consider_prior
+        consider_magic_clip = self._parameters.consider_magic_clip
+        sigmas0 = self._sigmas0
+        low = self._low[param_name]
+        high = self._high[param_name]
+        assert low is not None
+        assert high is not None
+
+        if n_observations == 0:
+            consider_prior = True
+
+        pairs_of_observation_and_idx = np.asarray(
+            [(low, -1)] + [(x, i) for i, x in enumerate(observations)] + [(high, n_observations)]
+        )
+        pairs_of_observation_and_idx = pairs_of_observation_and_idx[
+            np.argsort(pairs_of_observation_and_idx[:, 0])
+        ]
 
-        # We decide the weights.
-        unsorted_weights = weights_func(mus.size)
         if consider_prior:
-            sorted_weights = numpy.zeros_like(sorted_mus)
-            sorted_weights[:prior_pos] = unsorted_weights[order[:prior_pos]]
-            sorted_weights[prior_pos] = prior_weight
-            sorted_weights[prior_pos + 1 :] = unsorted_weights[order[prior_pos:]]
+
+            prior_mu = 0.5 * (low + high)
+            prior_sigma = 1.0 * (high - low)
+
+            mus = np.empty(n_observations + 1)
+            mus[:n_observations] = observations
+            mus[n_observations] = prior_mu
+
+            if sigmas0 is None:
```

Comment:
notes: Different implementation from multivariate_parzen_estimator.

Reply:
Might aid readability to leave a comment about this line, i.e. that it's true when we're not suggesting multivariate configurations and we don't need the heuristic.

Reply:
We could also do something along the following to improve locality 
```
if `use_simple_sigmas`:
    if self._sigmas0 is None:
        self._sigmas0 = self._precompute_sigmas0(observations)
# And, do not call `self._precompute_sigmas0` in `__init__`.
```